### ISSUE TITLE
Bug: [no-unnecessary-condition] False positive on let variable reassigned in awaited promise block

### ISSUE THREAD

**Issue by ericrav** [typescript-eslint/typescript-eslint#7813]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.9.5&fileType=.ts&code=IYZwngdgxgBAZgV2gFwJYHsI2QUxMgCjgBtgBzALhgCN11idgIBKGAbwFgAoGXmB5DTrEYAXnjBiIHAG5u3PjAD0S7AAtUIGFHQATHDE0wA8gGkKCvipgECrUQD52l3tb6o4NkuVa16Y7AAnBFkXZVUAX2Y7OS4w4AB3YFRBCBwEmAAFQPQAW00cAkC8egA3A0dnHkV3TyJSMl9hAORg0OqamGKQMsKAFgAmZljFKJG%2BeQ7DOr9iVk4pxR0IHoYAOmJ0MgIAclaQneGwiO4IoA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```typescript
async function test(flag: boolean) {
    let bool = false;

    await new Promise(resolve => {
        if (flag) bool = true;
        resolve(42);
    });
    
    // incorrectly reports bool as always false
    if (bool) {
        console.log('true');
    }
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

No ESLint warning should be reported because the value can be `true` or `false`

### Actual Result

ESLint reports value is always falsy

**Comment by Josh-Cena** [comment]
Hi, have you checked the docs? There's a mention of exactly this case, because so many people get confused by this. https://typescript-eslint.io/rules/no-unnecessary-condition/#when-not-to-use-it
